enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,826 @@
+	// the configured triggers to see if scaling up or down is required.
+	//
+	// When omitted, this means the user has no opinion and the value is left to
+	// the platform to choose a good default, which is subject to change over
+	// time. The current default is 30m.
+	SyncPeriod Duration
+	// ScaleUp allows the user to define specifics around scale up decisions.
+	// For example it allows the to specify possible triggers like CPU or memory
+	// utilization.
+	//
+	// When omitted, the value will default to disabling scaling up.
+	//
+	// +kubebuilder:validation:Optional
+	ScaleUp ScaleDefinition
+	// ScaleDown allows the user to define specifics around scale down
+	// decisions.
+	// For example it allows the to specify possible triggers like CPU or memory
+	// utilization.
+	//
+	// When omitted, the value will default to disabling scaling down.
+	//
+	// +kubebuilder:validation:Optional
+	ScaleDown ScalingDefinition
+}
+
+// MachineConfiguration defines the changes to be made to the
+// controlplanemachineset in case of scaling up or down. It lets the customer
+// specify platform specific options to increase or decrease CPU or memory.
+type MachineConfiguration struct {
+	// MachineType is the union discriminator.
+	// Users are expected to set this value to the name of the platform.
+	// Currently the only valid value is 'machines_v1beta1_machine_openshift_io'
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:="machines_v1beta1_machine_openshift_io"
+	// +kubebuilder:validation:Required
+	MachineType string
+	// OpenShiftMachineV1Beta1Machine defines the template for creating Machines
+	// from the v1beta1.machine.openshift.io API group.
+	OpenShiftMachinesV1Beta1Machine MachineConfigurationMAPI
+}
+
+type MachineConfigurationMAPI struct {
+	// Platform identifies the platform for which the FailureDomain represents.
+	// Currently supported values are AWS, Azure, and GCP.
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:="AWS","Azure","GCP","Nutanix","Openstack","Vsphere"
+	// +kubebuilder:validation:Required
+	Platform string
+	// AWS configures machine information for the AWS platform.
+	// +kubebuilder:validation:Optional
+	Aws AWSMachineConfiguration
+	// Azure configures machine information for the Azure platform.
+	// +kubebuilder:validation:Optional
+	Azure AzureMachineConfiguration
+	// Gcp configures machine information for the Gcp platform.
+	// +kubebuilder:validation:Optional
+	Gcp GCPMachineConfiguration
+	// Nutanix configures machine information for the Nutanix platform.
+	// +kubebuilder:validation:Optional
+	Nutanix NutanixMachineConfiguration
+	// Openstack configures machine information for the Openstack platform.
+	// +kubebuilder:validation:Optional
+	Openstack OpenstackMachineConfiguration
+	// Vsphere configures machine information for the Vsphere platform.
+	// +kubebuilder:validation:Optional
+	Vsphere VsphereMachineConfiguration
```

Comment:
Should these be types of lists of `*MachineConfiguration` types?

So `Aws []AWSMachineConfiguration`, etc? 

I'm trying to imagine what this would look like in it's current form in yaml, but I'm thinking something like this:

```yaml
OpenShiftMachinesV1Beta1Machine: 
    Platform: "AWS"
    Aws:
        - AWSMachineConfiguration:
            Weight: 0
            InstanceSize: "m5.2xlarge"
        - AWSMachineConfiguration:
            Weight: 1
            InstanceSize: "m5.4xlarge"
```

Am I understanding this correctly?